### PR TITLE
[SPARK-8237][SQL] Add misc function sha2

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-8237

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan can you help review this? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -44,7 +47,96 @@ case class Md5(child: Expression)
...
+/**
+ * A function that calculates the SHA-2 family of functions (SHA-224, SHA-256, SHA-384, and SHA-512)
+ * and returns it as a hex string. The first argument is the string or binary to be hashed. The
+ * second argument indicates the desired bit length of the result, which must have a value of 224,
+ * 256, 384, 512, or 0 (which is equivalent to 256). SHA-224 is supported starting from Java 8. If
+ * asking for an unsupported SHA function, the return value is NULL. If either argument is NULL or
+ * the hash length is not one of the permitted values, the return value is NULL.
+ */
+case class Sha2(left: Expression, right: Expression)
+  extends BinaryExpression with Serializable with ExpectsInputTypes { self: Product =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: remove `self: Product =>`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -1414,6 +1414,22 @@ object functions {
    */
   def md5(columnName: String): Column = md5(Column(columnName))
 
+  /**
+   * Calculates the SHA-2 family of hash functions and returns the value as a hex string.
+   *
+   * @group misc_funcs
+   * @since 1.5.0
+   */
+  def sha2(e: Column, bit: Int): Column = Sha2(e.expr, lit(bit).expr)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: bit -> numBits

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Should we check numBits is valid or not to fail fast?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Currently if given bit length is not one of the permitted values, the result value is NULL. So I think it should not be to fail fast?

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -363,6 +364,23 @@ def randn(seed=None):
     return Column(jc)
 
 
+@ignore_unicode_prefix
+@since(1.5)
+def sha2(col, length):
+    """Returns the hex string result of SHA-2 family of hash functions (SHA-224, SHA-256, SHA-384,
+       and SHA-512).
```
Comment (davies) [inline python/pyspark/sql/functions.py]: No indent this line.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: The handling of invalid bit length is useful for SQL (we can not easily valid the parameter in SQL parser), but for DataFrame API, it's good to fail fast, give better feedback to users.

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (davies) [inline python/pyspark/sql/functions.py]: Could you say a bit more what the length is?

**Comment by davies** [comment]
LGTM, left two minor comments (good to have).

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: +1 to fail fast

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: ok. I will add it later.

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8237][SQL] Add misc function sha2 (#6934)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Pull request closed without merging**